File: src/typescript/helpers/search/linear.ts
import { SearchResult } from "./types";

export function linearSearch(
  arr: number[],
  target: number,
  { loggingEnabled = false } = {}
): SearchResult {
  let steps = 0;

  for (let i = 0; i < arr.length; i++) {
    steps++;

    if (loggingEnabled) {
      console.log(`Step ${steps}: Check arr[${i}] = ${arr[i]}`);
    }

    if (arr[i] === target) {
      return { index: i, steps };
    }
  }

  return { index: -1, steps };
}

File: src/go/helpers/search/linear.go
package search

import "fmt"

func LinearSearch(values []int, target int, enableLogging bool) (int, int) {
	steps := 0
	for i, value := range values {
		steps++
		if enableLogging {
			fmt.Printf("  Step %d: Check arr[%d] = %d\n", steps, i, value)
		}

		if value == target {
			return i, steps
		}
	}
	return -1, steps
}
